### ISSUE TITLE
blip_sync panic/recovery handler

### ISSUE THREAD

**Issue by tleyden** [couchbase/sync_gateway#3255]
From https://github.com/couchbase/sync_gateway/issues/3249, in [these logs](https://gist.github.com/sridevi-15/8d7906d0001e04b1abec34edb93288f2#file-gistfile1-txt-L33-L47) 

It shows that this issue is crashing the entire sync gateway process.

Should there be high level panic/recovery handling for all blip handlers along the lines of the Go http library?

**Comment by adamcfraser** [comment]
Do we already have this issue for continuous changes over websocket today?  I don't see any recover() handling in that scenario, but I don't recall seeing any tickets where a continuous changes crash brought down SG.  

Also - any idea why the existing recover() in blipHandler.sendChanges() isn't catching this?

**Comment by tleyden** [comment]
> Also - any idea why the existing recover() in blipHandler.sendChanges() isn't catching this?

I think it's because of this goroutine being spawned:

```
// Returns the (ordered) union of all of the changes made to multiple channels.
func (db *Database) VectorMultiChangesFeed(chans base.Set, options ChangesOptions) (<-chan *ChangeEntry, error) {
	to := ""
	var userVbNo uint16
	if db.user != nil && db.user.Name() != "" {
		to = fmt.Sprintf("  (to %s)", db.user.Name())
		userVbNo = uint16(db.Bucket.VBHash(db.user.DocID()))
	}
	base.LogTo("Changes+", "Vector MultiChangesFeed(channels: %s, options: %+v) ... %s", chans, options, to)

	output := make(chan *ChangeEntry, 50)

	go func() {   // <---- spawn goroutine w/o panic handler
		var cumulativeClock *base.SyncSequenceClock
		var lastHashedValue string
		hashedEntryCount := 0
		defer func() {
			base.LogTo("Changes+", "MultiChangesFeed done %s", to)
			close(output)
		}()
```

Combined with the fact that [panic/recovery handling is per-goroutine](https://groups.google.com/forum/#!topic/golang-nuts/Bi43JRE-c6w)

**Comment by adamcfraser** [comment]
That explains why that recover isn't working, but leaves me wondering again why we haven't seen this more often (since both SimpleMultiChangesFeed and VectorMultiChangesFeed spawn their own goroutines).  I assume we've just shaken out the panic scenarios in SimpleMultiChangesFeed, and VectorMultiChangesFeed hasn't been used as extensively.

Back to the original question - I agree that we should add high-level panic/recovery to blip handlers.  And I think it's reasonable to enhance the goroutines in SimpleMultiChangesFeed and VectorMultiChangesFeed to do the same.

**Comment by tleyden** [comment]
>  I assume we've just shaken out the panic scenarios in SimpleMultiChangesFeed, and VectorMultiChangesFeed hasn't been used as extensively.

Yes, that would be my guess too.

> I agree that we should add high-level panic/recovery to blip handlers

This can't be as high level as I originally thought, unless we reduce the amount of goroutine spawning behind handler api calls, due to the per-goroutine panic/recovery handling.  Or, it can be, but it will be less effective and we'll still have to duplicate the panic/recovery handling deeper down the stack as you mentioned "enhance the goroutines in SimpleMultiChangesFeed and VectorMultiChangesFeed to do the same."

**Comment by adamcfraser** [comment]
Tasks:
- [x] Validate whether the top-level websocket server already handles panics in the main goroutine for the request
   - Force a panic in any handler prior to starting a new goroutine, check whether it kills the SG process
   - If this isn't caught, add recover handling to close the connection.
- [x] Validate that a panic in spawned goroutines in blip_sync will kill the SG process
  - [x] If so, add recover handling to the goroutines started by the following.  Ideally these should return error for the client request. 
     - handleSubChanges
     - sendBatchOfChanges
     - sendRevision
- [x] Check whether a panic in a spawned goroutine in the following message functions (in go-blip) will kill the SG process.  If so, add recover handling to these goroutines - may just need to re-raise the panic outside of the goroutine context to allow it to be caught by the handler-specific recovery. (needs clarification on scope of recover processing)
     - Message.nextFrameToSend
     - Message.asyncRead

**Comment by bbrks** [comment]
Confirmed that panics in above places do kill the SG process. 2 PRs above are to recover from these forced panics and keep the SG process running.

**Comment by bbrks** [comment]
Fixed in #3282 and https://github.com/couchbase/go-blip/issues/21
